client/client.go

```diff
@@ -0,0 +1,768 @@
+	// client package defined request hooks
+	builtinRequestHooks []RequestHook
+
+	// user defined response hooks
+	userResponseHooks []ResponseHook
+
+	// client package defined response hooks
+	builtinResponseHooks []ResponseHook
+
+	jsonMarshal   utils.JSONMarshal
+	jsonUnmarshal utils.JSONUnmarshal
+	xmlMarshal    utils.XMLMarshal
+	xmlUnmarshal  utils.XMLUnmarshal
+
+	cookieJar *CookieJar
+
+	// tls config
+	tlsConfig *tls.Config
+
+	// proxy
+	proxyURL string
+
+	// retry
+	retryConfig *RetryConfig
+}
+
+// R raise a request from the client.
+func (c *Client) R() *Request {
+	return AcquireRequest().SetClient(c)
+}
+
+// Request returns user-defined request hooks.
+func (c *Client) RequestHook() []RequestHook {
+	return c.userRequestHooks
+}
+
+// Add user-defined request hooks.
+func (c *Client) AddRequestHook(h ...RequestHook) *Client {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.userRequestHooks = append(c.userRequestHooks, h...)
+	return c
+}
+
+// ResponseHook return user-define response hooks.
+func (c *Client) ResponseHook() []ResponseHook {
+	return c.userResponseHooks
+}
+
+// Add user-defined response hooks.
+func (c *Client) AddResponseHook(h ...ResponseHook) *Client {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.userResponseHooks = append(c.userResponseHooks, h...)
+	return c
+}
+
+// JSONMarshal returns json marshal function in Core.
+func (c *Client) JSONMarshal() utils.JSONMarshal {
+	return c.jsonMarshal
+}
+
+// Set json encoder.
+func (c *Client) SetJSONMarshal(f utils.JSONMarshal) *Client {
+	c.jsonMarshal = f
+	return c
+}
+
+// JSONUnmarshal returns json unmarshal function in Core.
+func (c *Client) JSONUnmarshal() utils.JSONUnmarshal {
+	return c.jsonUnmarshal
+}
+
+// Set json decoder.
+func (c *Client) SetJSONUnmarshal(f utils.JSONUnmarshal) *Client {
+	c.jsonUnmarshal = f
+	return c
+}
+
+// XMLMarshal returns xml marshal function in Core.
+func (c *Client) XMLMarshal() utils.XMLMarshal {
+	return c.xmlMarshal
+}
+
+// Set xml encoder.
+func (c *Client) SetXMLMarshal(f utils.XMLMarshal) *Client {
+	c.xmlMarshal = f
+	return c
+}
+
+// XMLUnmarshal returns xml unmarshal function in Core.
+func (c *Client) XMLUnmarshal() utils.XMLUnmarshal {
+	return c.xmlUnmarshal
+}
+
+// Set xml decoder.
+func (c *Client) SetXMLUnmarshal(f utils.XMLUnmarshal) *Client {
+	c.xmlUnmarshal = f
+	return c
+}
+
+// TLSConfig returns tlsConfig in client.
+// If client don't have tlsConfig, this function will init it.
+func (c *Client) TLSConfig() *tls.Config {
+	if c.tlsConfig == nil {
+		c.tlsConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+		}
+	}
+
+	return c.tlsConfig
+}
+
+// SetTLSConfig sets tlsConfig in client.
+func (c *Client) SetTLSConfig(config *tls.Config) *Client {
+	c.tlsConfig = config
+	return c
+}
+
+// SetCertificates method sets client certificates into client.
+func (c *Client) SetCertificates(certs ...tls.Certificate) *Client {
+	config := c.TLSConfig()
+	config.Certificates = append(config.Certificates, certs...)
```

Comment:
Does this enforce ClientCert Auth? We should add an option for setting ClientAuth type https://pkg.go.dev/crypto/tls#ClientAuthType

Reply:
> Does this enforce ClientCert Auth? We should add an option for setting ClientAuth type https://pkg.go.dev/crypto/tls#ClientAuthType

It can be set via `TLSConfig()` or `SetTLSConfig(config *tls.Config)`